### PR TITLE
analytical_gradient

### PR DESCRIPTION
**Pull request by jonathanward**

### CODE REVIEW

**Comment by ncrubin** [comment]
I will continue to review for correctness but general things,

follow pep8
remove linting hints
more documentation for grove-docs.

File: grove/pyqaoa/analytical_gradient.py [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+
+"""
+This module implements the analytical gradient as defined in the paper:
+'Practical Optimization for hybrid quantum-classical algorithms'
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: arXiv:1701.01450 [quant-ph]

File: grove/pyqaoa/analytical_gradient.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+"""
+
+import pyquil.quil as pq
+from pyquil.gates import * # pylint: disable=unused-wildcard-import,wildcard-import
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: The wildcard import is currently standard for gate imports.

File: grove/pyqaoa/analytical_gradient.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+from pyquil.paulis import * # pylint: disable=unused-wildcard-import,wildcard-import,redefined-builtin
+
+def generate_make_controlled(ancilla_qubit_index):
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: :+1:

File: grove/pyqaoa/analytical_gradient.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    """
+    Creates a function which maps operators to controlled operators
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: :+1:

File: grove/pyqaoa/analytical_gradient.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    Generates the analytical gradient corresponding to a list of hamiltonians
+    :param (list[pq.Program]) hamiltonians: these must be in the correct order
+    :param (PauliSum) cost_hamiltonian: for the expectation values
+    :param (int) steps: the number of levels in the trotterization
+    :param (int) num_qubits: the number of qubits allocated
+    :return (function) gradient: expectation value of the analytical_gradient
+    """
+    p_unitaries = [exponential_map_hamiltonian(hamiltonian)
+                   for hamiltonian in hamiltonians]
+    def gradient(steps_parameters): # pylint: disable=too-many-locals
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: :+1:

File: grove/pyqaoa/analytical_gradient.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    :param (PauliSum) hamiltonian: a hermitian hamiltonian
+    :param (float) parameter: the coefficient in the exponential
+    :return (function) p_unitary: the generated parameterized unitary
+    """
+    p_unitary_list = []
+    for term in hamiltonian:
+        p_unitary_term = exponential_map(term)
+        p_unitary_list.append(p_unitary_term)
+
+    def p_unitary(parameter):
```
Comment (jonathanward) [inline grove/pyqaoa/analytical_gradient.py]: which one?

**Comment by ncrubin** [comment]
This will be further reviewed an refactored for a PR into the VQE module.

**Pull request closed without merging**